File: src/transformers/models/qwen3/Qwen3DecoderLayer.cpp
#include "transformers/models/qwen3/Qwen3DecoderLayer.hpp"
#include "transformers/models/qwen3/Qwen3Config.hpp"
#include "transformers/models/qwen3/Qwen3RMSNorm.hpp"
#include "transformers/models/qwen3/Qwen3MLP.hpp"
#include "transformers/models/qwen3/Qwen3Attention.hpp"
#include "nn/attention/FlashAttentionOp.hpp"

Qwen3DecoderLayer::Qwen3DecoderLayer(const Qwen3Config& config, int layer_idx) {        
    modules["self_attn"] = std::make_shared<Qwen3Attention<FlashAttentionOp>>(config, layer_idx);
    modules["mlp"] = std::make_shared<Qwen3MLP>(config);
    modules["input_layernorm"] = std::make_shared<Qwen3RMSNorm>(config.hidden_size, config.rms_norm_eps);
    modules["post_attention_layernorm"] = std::make_shared<Qwen3RMSNorm>(config.hidden_size, config.rms_norm_eps);
}

Tensor Qwen3DecoderLayer::forward(
    Runtime& runtime,
    Qwen3RotaryEmbedding& rotary_emb,
    Tensor hidden_states,
    Tensor position_ids, 
    std::optional<Tensor> attention_mask,
    std::optional<Tensor> past_key_values,
    std::optional<bool> use_cache
) {
    
    // Self Attention Block
    auto residual = hidden_states;
    
    auto input_layernorm = std::static_pointer_cast<Qwen3RMSNorm>(modules["input_layernorm"]);
    hidden_states = input_layernorm->forward(runtime, hidden_states);

    auto self_attn = std::static_pointer_cast<Qwen3Attention<FlashAttentionOp>>(modules["self_attn"]);
    hidden_states = self_attn->forward(runtime, rotary_emb, hidden_states, position_ids, attention_mask, past_key_values);

    hidden_states = residual + hidden_states;

    // Fully Connected Block
    residual = hidden_states;
    
    auto post_attention_layernorm = std::static_pointer_cast<Qwen3RMSNorm>(modules["post_attention_layernorm"]);
    hidden_states = post_attention_layernorm->forward(runtime, hidden_states);

    auto mlp = std::static_pointer_cast<Qwen3MLP>(modules["mlp"]);
    hidden_states = mlp->forward(runtime, hidden_states);

    hidden_states = residual + hidden_states;

    return hidden_states;
}

File: test/src/transformers/models/qwen3/Qwen3DecoderLayer.py
from utils import TestCase
import torch
from torch.nn.attention import SDPBackend, sdpa_kernel
from transformers.models.qwen3.modeling_qwen3 import Qwen3Config, Qwen3DecoderLayer, Qwen3RotaryEmbedding

class TestTransformersQwen3DecoderLayer(TestCase):
    def test(self):
        config = Qwen3Config(
            hidden_size=8,
            intermediate_size=16,
            num_attention_heads=2,
            num_key_value_heads=2,
            max_position_embeddings=32,
            _attn_implementation="sdpa"
        )

        model = Qwen3DecoderLayer(config, layer_idx=0)
        rotary_emb = Qwen3RotaryEmbedding(config)

        model.self_attn.is_causal = False

        hidden_states = torch.randn(1, 4, 8)
        position_ids = torch.arange(4).unsqueeze(0)

        position_embeddings = rotary_emb(
            hidden_states,
            position_ids,
        )

        with sdpa_kernel(SDPBackend.FLASH_ATTENTION):
            expected = model(
                hidden_states=hidden_states,
                position_embeddings=position_embeddings,
                attention_mask=None,
            )

        actual = self.cli(
            "Qwen3DecoderLayer",
            "--hidden_size", "8",
            "--intermediate_size", "16",
            "--num_attention_heads", "2",
            "--num_key_value_heads", "2",
            "--max_position_embeddings", "32",
            "--layer_idx", "0",
            "--hidden_states", str(hidden_states.tolist()),
            "--position_ids", str(position_ids.tolist()),
            *self.params(model)
        )

        self.assertTensors(actual, [expected])
